File: Hackerrank/Algorithms/Graph Theory/Roads and libraries.py
#!/bin/python3
import queue
def roadsAndLibraries(n, c_lib, c_road, cities):
    # Complete this function
    if c_road >= c_lib:
        return c_lib * n
    else:
        comp = 0
        visited = [False]*(n+1)
        for c in range(1,n+1):
            # print(c)
            if(not visited[c]):
                comp = comp + 1
                q = queue.Queue(maxsize = 10**5)
                q.put(c)
                visited[c] = True
                while(not q.empty()):
                    w = q.get()
                    l = [x for x in cities if w in x]
                    l1 = [x[0] for x in l if x[1] == w] + [x[1] for x in l if x[0] == w]
                    for x in l1:
                        if not visited[x]:
                            visited[x] = True
                            q.put(x)
                        
        return len(cities)*c_road + (c_lib-c_road)* comp
if __name__ == "__main__":
    q = int(input().strip())
    for a0 in range(q):
        n, m, c_lib, c_road = input().strip().split(' ')
        n, m, c_lib, c_road = [int(n), int(m), int(c_lib), int(c_road)]
        cities = []
        for cities_i in range(m):
           cities_t = [int(cities_temp) for cities_temp in input().strip().split(' ')]
           cities.append(cities_t)
        result = roadsAndLibraries(n, c_lib, c_road, cities)
        print(result)

File: Hackerrank/Algorithms/Graph Theory/Roads and libraries.cpp
#include <cmath>
#include <cstdio>
#include <cstring>
#include <vector>
#include <iostream>
#include <algorithm>
#include <queue>
using namespace std;


int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
int q;
cin >> q;
while(q--){
	long int n,m,c_road,c_lib;
	cin >> n >> m >> c_lib>> c_road;
	vector <int> adj[n];
	for(int i = 0 ; i<m ; i++){
		int x,y;
		cin >> x >> y;
		adj[x-1].push_back(y);
		adj[y-1].push_back(x);
		}

	if(c_lib <= c_road)
		cout << n * c_lib << endl;
	else{
		int visited[n+1];
		memset(visited,0,sizeof(visited));
		int comp = 0;
		for(int i = 1 ; i <= n ; i++){
			if(!visited[i]){
				comp++;
				queue<long int> q;
				q.push(i);
				visited[i] = true;
				while(!q.empty()){
					int k = q.front();
					q.pop();
					vector<int> v = adj[k-1];
					for(int j=0 ; j<v.size() ; j++){
						if(!visited[v[j]]){
							visited[v[j]] = true;
							q.push(v[j]);
						}
					}
				}
			}
		}
		//cout << "Comp: " << comp << "n,m: "<< n<<" " << m <<endl;
		long long int out = n*c_road + comp*(c_lib-c_road) ;
		cout << out << endl;
	}

}

}



